hermes-client/src/main/java/pl/allegro/tech/hermes/client/ReactiveHermesClient.java

```diff
@@ -0,0 +1,324 @@
+                .onErrorResume(this::mapExceptionToFailedAttempt)
+                .subscribeOn(scheduler);
+
+        return retry(message, sendOnceResult)
+                .doOnSuccess(hr -> {
+                    if (hr.response.isSuccess()) {
+                        handleSuccessfulRetry(message, hr.getAttempt());
+                    } else {
+                        handleFailure(message, hr.getAttempt());
+                    }
+                })
+                .map(result -> result.response)
+                .doFinally(s -> currentlySending.decrementAndGet());
+    }
+
+    private Mono<Attempt> retry(HermesMessage message, Mono<Result> sendOnceResult) {
+        Retry retrySpec = prepareRetrySpec(message);
+        return sendOnceResult
+                .flatMap(this::unwrapFailedAttemptAsException)
+                .retryWhen(retrySpec)
+                .subscriberContext(ctx -> ctx.put(RETRY_CONTEXT_KEY, HermesRetryContext.emptyRetryContext()))
+                .onErrorResume(Exception.class, exception -> {
+                    if (isRetryExhausted(exception)) {
+                        RetryFailedException rfe = (RetryFailedException) (exception.getCause());
+                        handleMaxRetriesExceeded(message, rfe.attempt);
+                        Throwable cause = rfe.getCause();
+                        if (cause instanceof ShouldRetryException) {
+                            ShouldRetryException sre = (ShouldRetryException) cause;
+                            return Mono.just((Attempt) Result.attempt(sre.hermesResponse, rfe.attempt));
+                        }
+                        handleFailure(message, rfe.getAttempt());
+                    }
+                    return Mono.error(exception);
+                });
+    }
+
+    private Mono<Attempt> unwrapFailedAttemptAsException(Result result) {
+        if (result instanceof Failed) {
+            Failed failed = (Failed) result;
+            return Mono.error(new RetryFailedException(failed.attempt, failed.cause));
+        } else {
+            return Mono.just((Attempt) result);
+        }
+    }
+
+    private Mono<Result> mapExceptionToFailedAttempt(Throwable throwable) {
+        return getNextAttempt().map(attempt -> Result.failure(throwable, attempt));
+    }
+
+    private Mono<Result> testRetryCondition(HermesResponse response) {
+        return getNextAttempt()
+                .map(attempt -> {
+                    if (retryCondition.test(response)) {
+                        return Result.failure(new ShouldRetryException(response), attempt);
+                    } else {
+                        return Result.attempt(response, attempt);
+                    }
+                });
+    }
+
+    private Retry prepareRetrySpec(HermesMessage message) {
+        if (!retrySleep.isZero()) {
+            return Retry.max(maxRetries)
+                    .doAfterRetry(signal -> handleFailedAttempt(message, signal.totalRetries() + 1));
```

Comment:
`doAfterRetry` requires the signal consumer to be a non-blocking function. `MessageDeliveryListener`'s methods can be blocking so I think it's a good idea to introduce `ReactiveMessageDeliveryListener` with a similar interface as `MessageDeliveryListener` but with non-blocking methods (returning `Mono<Void>`).

The default, one listener gets injected into `ReactiveHermesClient` and it reports metrics to meter registry. I'm not sure whether the `MeterRegistry` implementation is blocking or not. If it is not blocking then the implementation of `ReactiveMessageDeliveryListener` will be pretty much straight forward.